Add description for the append constraint in xpath constraint for access rule (#4363)
mendix/docs: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md

### 2.4 XPath Constraint Tab {#xpath-constraint}

An [XPath constraint](/refguide/xpath-constraints/) can be used to constrain the set of objects to which the access rule applies. If the constraint rule is true, the rule applies to that object. If the XPath constraint is empty, the rule applies to all objects of the entity.

{{< figure src="/attachments/refguide/modeling/domain-model/entities/access-rules/access-rule-xpath-tab.png" >}}

{{% alert color="warning" %}}
XPath constraints can only be applied to persistable entities as they are applied by the database. Defining XPath constraints for non-persistable entities results in consistency errors.
{{% /alert %}}

There are two constraints that can be appended easily with a single button click. 

#### 2.4.1 Owner

> The **Owner** button adds an XPath constraint so the access rule is only applied to the current user.

Reviewer: Would it be clearer to say:
```suggestion
The **Owner** button adds an XPath constraint so the access rule is only applied if the object owner is the current user.
```

Author: Agree.

```java {linenos=false}
[System.owner='[%CurrentUser%]']
```

This constraint is only valid when the [Store 'owner'](/refguide/entities/#store-owner) checkbox in the **System members** section of the entity properties is checked.

#### 2.4.2 Path to User

> The **Path to user...** button adds an XPath constraint so the access rule is only applied when the associated object (directly or indirectly) is the current user. When you click **Path to user...**, you can select a path to an associated entity that is either a `System.User` or a specialization of `System.User`. This is then converted into an XPath constraint for the access rule.

Reviewer: Perhaps clarify that it is a *User* object?
```suggestion
The **Path to user...** button adds an XPath constraint so the access rule is only applied when the User object which is associated (directly or indirectly) is the current user. When you click **Path to user...**, you can select a path to an associated entity that is either a `System.User` or a specialization of `System.User`. This is then converted into an XPath constraint for the access rule.
```

Author: Done.

```java {linenos=false}
[Module.Order_Customer = '[%CurrentUser%]']
```

As an example:

> 1. Assume that the **Customer** entity is a specialization of the **User** entity. The **Order** entity is associated to the **Customer** entity via the **Order_Customer** association.

Reviewer: ```suggestion
1. Assume that the **Customer** entity is a specialization of the **User** entity. The **Order** entity is associated with the **Customer** entity via the **Order_Customer** association.
```

Author: Done.

{{% /alert %}}

There are two constraints that can be appended easily with a single button click. 

#### 2.4.1 Owner

The **Owner** button adds an XPath constraint so the access rule is only applied to the current user.

```java {linenos=false}
[System.owner='[%CurrentUser%]']
```

This constraint is only valid when the [Store 'owner'](/refguide/entities/#store-owner) checkbox in the **System members** section of the entity properties is checked.

#### 2.4.2 Path to User

The **Path to user...** button allows you to create an XPath constraint to check that the entity is associated (directly or indirectly) with the current user. When you click **Path to user...**, you can select a path to an entity that is a `System.User` or a specialization of `System.User`. This is converted into an XPath constraint for the access rule. This means that the rule is only applied when the user path you selected is associated with the current user.

#### 2.4.3 Example

Assume that the **Customer** entity is a specialization of the **User** entity. The **Order** entity is associated to the **Customer** entity via the **Order_Customer** association.

> A logged-in customer is allowed to view personal orders, but is not allowed to view the orders of other customers. This is accomplished by using the following XPath constraint in the access rule of the **Order** entity:

Reviewer: @paulmendix - did you mean something like:
```suggestion
A logged-in customer is allowed to view personal orders, but is not allowed to view the orders of other customers. This is accomplished by using the following XPath constraint in the access rule of the **Order** entity, which you can add using the **Path to user...** button:
```

Author: I think I have it now. Can you check if it is more readable?

Reviewer 2: Looks good to me

{{< figure src="/attachments/refguide/modeling/domain-model/entities/access-rules/access-rule-xpath-tab.png" >}}

> 

Reviewer 2: This is definitely not your fault, as you are trying to improve this documentation, but I find reading this existing section super difficult. To me it feels really chaotic.

Adding more information here makes it extra complex. 

# Proposal

I propose we add two separate sections:

`#### 2.4.1 Owner` and `#### 2.4.2 Path to user...`

Then the Owner section can be ⬇

---

The  **Owner** button adds an XPath constraint so the access rule is only applied to the current user. 

This constraint is only valid when the **Store 'owner'** checkbox in the **System members** section is checked.

![image](https://user-images.githubusercontent.com/36693526/160430438-f4c57817-2203-4ea8-bdf1-e304e43d512d.png)

---

Note that I initially really liked that you added the actual XPath constraint, but when I see it as a user I want to fully understand it, and then we get to hidden one-way navigable associations, so then I changed my mind. Maybe better to not show it.

---

Then the 'Path to user... ' section can be ⬇

---

The **Path to user...** button is useful to create an XPath constraint to check if the entity is (indirectly) connected to the current user. When you press **Path to user...** you can select a path to an entity that is derived from **System.User** and converts that into an XPath constraint for the access rule. This means the access rule is only applied when the user path you selected is equal to the current user.

---

Note that you mention `We provided 2 helper buttons`. I don't remember seeing `We` in the documentation.

Author: Done.

Reviewer 2: Looks good to me

{{< figure src="/attachments/refguide/modeling/domain-model/entities/access-rules/access-rule-xpath-tab.png" >}}

#### 2.4.1 Owner

The **Owner** button adds an XPath constraint so the access rule is only applied to the current user.

```java
[System.owner='[%CurrentUser%]']
```

This constraint is only valid when the [Store 'owner'](/refguide/entities/#343-store-owner) checkbox in the System members section is checked.

#### 2.4.2

The **Path to user...** button is useful to create an XPath constraint to check if the entity is (indirectly) associated to the current user. When you press **Path to user...** you can select a path to an entity that is a `System.User` or a specialization of `System.User` and converts that into an XPath constraint for the access rule. This means the access rule is only applied when the user path you selected is equal to the current user.

> For example, the **Customer** entity is a specialization of the **User** entity. The **Order** entity is associated to the **Customer** entity.

Reviewer 2: I think this example belongs to section 2.4 and not to 2.4.2

Author: Yes it is.

Reviewer 2: Then I think the sentence `This is accomplished by using the following XPath constraint in the access rule of the **Order** entity:` needs to be changed so you describe how to use the 'Path to user' button. Now it gives the impression you have to write the XPath constraint yourself.

Author: Indeed I had an attempt to refactor "For example, ..." but I ended up with more complex explanation. I choose to revert back to how it was without realizing that I re-introduced the old sentences. I will make another attempt to rephrase the example part.

Author: I think I have it now. Can you check?

Reviewer 2: Looks good to me
